**10.文件和异常**

- 从文件中读取数据

  文本文件可存储的数据量多得难以置信：天气数据、交通数据、社会经济数据、文学作品等。每当需要分析或修改存储在文件中的信息时，读取文件都很有用，对数据分析应用程序来说尤其如此。例如，你可以编写一个这样的程序：读取一个文本文件的内容，重新设置这些数据的格式并将其写入文件，让浏览器能够显示这些内容。要使用文本文件中的信息，首先需要将信息读取到内存中。为此，你可以一次性读取文件的全部内容，也可以以每次一行的方式逐步读取。

  1. 读取整个文件

     要读取文件，需要一个包含几行文本的文件。下面首先来创建一个文件，它包含精确到小数点后 30 位的圆周率值，且在小数点后每 10 位处都换行：

     ```python
     with open('1.txt') as file_object:
         contents = file_object.read()
         print(contents)
         
     3.1415926535
     8979323846
     2643383279
     ```

     在这个程序中，第 1 行代码做了大量的工作。我们先来看看函数 open() 。要以任何方式使用文件 —— 哪怕仅仅是打印其内容，都得先 打开 文件，这样才能访问它。函数 open()接受一个参数：要打开的文件的名称。 Python 在当前执行的文件所在的目录中查找指定的文件。在这个示例中，当前运行的是 test.py ，因此 Python 在 test.py 所在的目录中查找 1.txt 。函数 open() 返回一个表示文件的对象。在这里， open('1.txt') 返回一个表示文件 1.txt 的对象； Python 将这个对象存储在我们将在后面使用的变量中。关键字 with 在不再需要访问文件后将其关闭。在这个程序中，注意到我们调用了 open() ，但没有调用 close() ；你也可以调用 open() 和 close() 来打开和关闭文件，但这样做时，如果程序存在 bug ，导致 close() 语句未执行，文件将不会关闭。这看似微不足道，但未妥善地关闭文件可能会导致数据丢失或受损。如果在程序中过早地调用 close() ，你会发现需要使用文件时它已 关闭 （无法访问），这会导致更多的错误。并非在任何情况下都能轻松确定关闭文件的恰当时机，但通过使用前面所示的结构，可让 Python 去确定：你只管打开文件，并在需要时使用它， Python 自会在合适的时候自动将其关闭

  2. 文件路径

     当你将类似 1.txt 这样的简单文件名传递给函数 open() 时， Python 将在当前执行的文件（即 .py 程序文件）所在的目录中查找文件。根据你组织文件的方式，有时可能要打开不在程序文件所属目录中的文件。例如，你可能将程序文件存储在了文件夹 python_work 中，而在文件夹 python_work 中，有一个名为text_files 的文件夹，用于存储程序文件操作的文本文件。虽然文件夹 text_files 包含在文件夹 python_work 中，但仅向 open() 传递位于该文件夹中的文件的名称也不可行，因为 Python只在文件夹 python_work 中查找，而不会在其子文件夹 text_files 中查找。要让 Python 打开不与程序文件位于同一个目录中的文件，需要提供 文件路径 ，它让 Python 到系统的特定位置去查找

  3. 逐行读取

     读取文件时，常常需要检查其中的每一行：你可能要在文件中查找特定的信息，或者要以某种方式修改文件中的文本。例如，你可能要遍历一个包含天气数据的文件，并使用天气描述中包含字样 sunny 的行。在新闻报道中，你可能会查找包含标签 <headline> 的行，并按特定的格式设置它。要以每次一行的方式检查文件，可对文件对象使用 for 循环：

     ```python
     filename = '1.txt'
     with open(filename) as file_object:
     	for line in file_object:
     	print(line)
         
     3.1415926535
     
     8979323846
     
     2643383279
     ```

     为何会出现这些空白行呢？因为在这个文件中，每行的末尾都有一个看不见的换行符，而 print 语句也会加上一个换行符，因此每行末尾都有两个换行符：一个来自文件，另一个来自 print 语句。要消除这些多余的空白行，可在 print 语句中使用 rstrip() ：

     ```python
     filename = '1.txt'
     with open(filename) as file_object:
     	for line in file_object:
     	print(line.rstrip())
         
     3.1415926535
     8979323846
     2643383279
     ```

  4. 创建一个包含文件各行内容的列表

     使用关键字 with 时， open() 返回的文件对象只在 with 代码块内可用。如果要在 with 代码块外访问文件的内容，可在 with 代码块内将文件的各行存储在一个列表中，并在 with 代码块外使用该列表：你可以立即处理文件的各个部分，也可推迟到程序后面再处理。下面的示例在 with 代码块中将文件1.txt 的各行存储在一个列表中，再在 with 代码块外打印它们

     ```python
     filename = '1.txt'
     with open(filename) as file_object:
         lines = file_object.readlines()
     for line in lines:
         print(line.rstrip())
     ```

  5. 使用文件的内容

     将文件读取到内存中后，就可以以任何方式使用这些数据了。下面以简单的方式使用圆周率的值。首先，我们将创建一个字符串，它包含文件中存储的所有数字，且没有任何空格：

     ```python
     filename = '1.txt'
     with open(filename) as file_object:
         lines = file_object.readlines()
     pi_string = ''
     for line in lines:
         pi_string += line.rstrip()
     print(pi_string)
     print(len(pi_string))
     
     3.141592653589793238462643383279
     32
     ```

- 写入文件

  保存数据的最简单的方式之一是将其写入到文件中。通过将输出写入文件，即便关闭包含程序输出的终端窗口，这些输出也依然存在：你可以在程序结束运行后查看这些输出，可与别人分享输出文件，还可编写程序来将这些输出读取到内存中并进行处理

  1. 写入空文件

     要将文本写入文件，你在调用 open() 时需要提供另一个实参，告诉 Python 你要写入打开的文件。为明白其中的工作原理，我们来将一条简单的消息存储到文件中，而不是将其打印到屏幕上：

     ```python
     filename = '1.txt'
     with open(filename, 'w') as file_object:
         file_object.write("I love programming.")
     ```

  2. 写入多行

     函数 write() 不会在你写入的文本末尾添加换行符，因此如果你写入多行时没有指定换行符，文件看起来可能不是你希望的那样，要让每个字符串都单独占一行，需要在 write() 语句中包含换行符

     ```python
     filename = '1.txt'
     with open(filename, 'w') as file_object:
         file_object.write("I love programming.\n")
         file_object.write("I love creating new games.\n")
     ```

  3. 附加到文件

     如果你要给文件添加内容，而不是覆盖原有的内容，可以 附加模式 打开文件。你以附加模式打开文件时， Python 不会在返回文件对象前清空文件，而你写入到文件的行都将添加到文件末尾。如果指定的文件不存在， Python 将为你创建一个空文件。下面来修改 test.py ，在既有文件 1.txt 中再添加一些你酷爱编程的原因：

     ```python
     filename = '1.txt'
     with open(filename, 'a') as file_object:
         file_object.write("I also love finding meaning in large datasets.\n")
         file_object.write("I love creating apps that can run in a browser.\n")
     ```

- 异常

  Python 使用被称为 异常 的特殊对象来管理程序执行期间发生的错误。每当发生让 Python 不知所措的错误时，它都会创建一个异常对象。如果你编写了处理该异常的代码，程序将继续运行；如果你未对异常进行处理，程序将停止，并显示一个 traceback ，其中包含有关异常的报告。异常是使用 try-except 代码块处理的。 try-except 代码块让 Python 执行指定的操作，同时告诉 Python 发生异常时怎么办。使用了 try-except 代码块时，即便出现异常，程序也将继续运行：显示你编写的友好的错误消息，而不是令用户迷惑的 traceback 。

  1. 处理 ZeroDivisionError  异常

     面来看一种导致 Python 引发异常的简单错误。你可能知道不能将一个数字除以 0 ，但我们还是让 Python 这样做吧

     ```python
     print(5/0)
     
     Traceback (most recent call last):
       File "E:/pythonwork/test.py", line 1, in <module>
         print(5/0)
     ZeroDivisionError: division by zero
     ```

     在上述 traceback 中，指出的错误 ZeroDivisionError 是一个异常对象。 Python 无法按你的要求做时，就会创建这种对象。在这种情况下， Python 将停止运行程序，并指出引发了哪种异常，而我们可根据这些信息对程序进行修改。下面我们将告诉 Python ，发生这种错误时怎么办；这样，如果再次发生这样的错误，我们就有备无患了。

  2. 使用 try-except  代码块

     当你认为可能发生了错误时，可编写一个 try-except 代码块来处理可能引发的异常。你让 Python 尝试运行一些代码，并告诉它如果这些代码引发了指定的异常，该怎么办。处理 ZeroDivisionError 异常的 try-except 代码块类似于下面这样

     ```python
     try:
         print(5/0)
     except ZeroDivisionError:
         print("You can't divide by zero!")
         
     You can't divide by zero!
     ```

     如果 try-except 代码块后面还有其他代码，程序将接着运行，因为已经告诉了 Python 如何处理这种错误

  3. 使用异常避免崩溃

     发生错误时，如果程序还有工作没有完成，妥善地处理错误就尤其重要。这种情况经常会出现在要求用户提供输入的程序中；如果程序能够妥善地处理无效输入，就能再提示用户提供有效输入，而不至于崩溃。下面来创建一个只执行除法运算的简单计算器：

     ```python
     print("Give me two numbers, and I'll divide them.")
     print("Enter 'q' to quit.")
     while True:
         first_number = input("\nFirst number: ")
         if first_number == 'q':
             break
         second_number = input("Second number: ")
         if second_number == 'q':
             break
         answer = int(first_number) / int(second_number)
         print(answer)
         
     Give me two numbers, and I'll divide them.
     Enter 'q' to quit.
     
     First number: 5
     Second number: 0
     Traceback (most recent call last):
       File "E:/pythonwork/test.py", line 10, in <module>
         answer = int(first_number) / int(second_number)
     ZeroDivisionError: division by zero
     ```

     程序崩溃可不好，但让用户看到 traceback 也不是好主意。不懂技术的用户会被它们搞糊涂，而且如果用户怀有恶意，他会通过 traceback 获悉你不希望他知道的信息。例如，他将知道你的程序文件的名称，还将看到部分不能正确运行的代码。有时候，训练有素的攻击者可根据这些信息判断出可对你的代码发起什么样的攻击

  4. else  代码块

     通过将可能引发错误的代码放在 try-except 代码块中，可提高这个程序抵御错误的能力。错误是执行除法运算的代码行导致的，因此我们需要将它放到 try-except 代码块中。这个示例还包含一个 else 代码块；依赖于 try 代码块成功执行的代码都应放到 else 代码块中

     ```python
     print("Give me two numbers, and I'll divide them.")
     print("Enter 'q' to quit.")
     while True:
         first_number = input("\nFirst number: ")
         if first_number == 'q':
             break
         second_number = input("Second number: ")
         try:
             answer = int(first_number) / int(second_number)
         except ZeroDivisionError:
             print("You can't divide by 0!")
         else:
             print(answer)
             
     Give me two numbers, and I'll divide them.
     Enter 'q' to quit.
     
     First number: 5
     Second number: 0
     You can't divide by 0!
     
     First number: 5
     Second number: 2
     2.5
     ```

     try-except-else 代码块的工作原理大致如下： Python 尝试执行 try 代码块中的代码；只有可能引发异常的代码才需要放在 try 语句中。有时候，有一些仅在 try 代码块成功执行时才需要运行的代码；这些代码应放在 else 代码块中。 except 代码块告诉 Python ，如果它尝试运行 try 代码块中的代码时引发了指定的异常，该怎么办。通过预测可能发生错误的代码，可编写健壮的程序，它们即便面临无效数据或缺少资源，也能继续运行，从而能够抵御无意的用户错误和恶意的攻击

  5. 处理 FileNotFoundError  异常

     使用文件时，一种常见的问题是找不到文件：你要查找的文件可能在其他地方、文件名可能不正确或者这个文件根本就不存在。对于所有这些情形，都可使用 try-except 代码块以直观的方式进行处理。我们来尝试读取一个不存在的文件。下面的程序尝试读取文件 1.txt 的内容，但我没有将这个文件存储在 1.py 所在的目录中：

     ```python
     filename = '1.txt'
     with open(filename) as f_obj:
     	contents = f_obj.read()
         
     Traceback (most recent call last):
     File "1.py", line 3, in <module>
     with open(filename) as f_obj:
     FileNotFoundError: [Errno 2] No such file or directory: '1.txt'
     ```

     在上述 traceback 中，最后一行报告了 FileNotFoundError 异常，这是 Python 找不到要打开的文件时创建的异常。在这个示例中，这个错误是函数 open() 导致的，因此要处理这个错误，必须将 try 语句放在包含 open() 的代码行之前：

     ```python
     filename = '1.txt'
     try:
     	with open(filename) as f_obj:
     		contents = f_obj.read()
     except FileNotFoundError:
     	msg = "Sorry, the file " + filename + " does not exist."
     	print(msg)
     ```

  6. 分析文本

     方法 split() 以空格为分隔符将字符串分拆成多个部分，并将这些部分都存储到一个列表中。结果是一个包含字符串中所有单词的列表，虽然有些单词可能包含标点

     ```python
     filename = '1.txt'
     try:
         with open(filename) as f_obj:
             contents = f_obj.read()
     except FileNotFoundError:
         msg = "Sorry, the file " + filename + " does not exist."
         print(msg)
     else:
         #  计算文件大致包含多少个单词
         words = contents.split()
         num_words = len(words)
         print("The file " + filename + " has about " + str(num_words) + " words.")
         
     The file 1.txt has about 26 words.
     ```

  7. 使用多个文件

     下面多分析几个文件。这样做之前，我们先将这个程序的大部分代码移到一个名为 count_words() 的函数中，这样对多个文件进行分析时将更容易，这些代码大都与原来一样，我们只是将它们移到了函数 count_words() 中，并增加了缩进量。修改程序的同时更新注释是个不错的习惯，因此我们将注释改成了文档字符串，并稍微调整了一下措辞，现在可以编写一个简单的循环，计算要分析的任何文本包含多少个单词了。为此，我们将要分析的文件的名称存储在一个列表中，然后对列表中的每个文件都调用 count_words()

     ```python
     def count_words(filename):
         """ 计算一个文件大致包含多少个单词 """
         try:
             with open(filename) as f_obj:
                 contents = f_obj.read()
         except FileNotFoundError:
             msg = "Sorry, the file " + filename + " does not exist."
             print(msg)
         else:
             #  计算文件大致包含多少个单词
             words = contents.split()
             num_words = len(words)
             print("The file " + filename + " has about " + str(num_words) +
                 " words.")
     filenames = ['1.txt', '2.txt','3.txt']
     for filename in filenames:
         count_words(filename)
         
     The file 1.txt has about 26 words.
     The file 2.txt has about 15 words.
     Sorry, the file 3.txt does not exist.
     ```

     文件 3.txt 不存在，但这丝毫不影响这个程序处理其他文件

  8. 失败时一声不吭

     在前一个示例中，我们告诉用户有一个文件找不到。但并非每次捕获到异常时都需要告诉用户，有时候你希望程序在发生异常时一声不吭，就像什么都没有发生一样继续运行。要让程序在失败时一声不吭，可像通常那样编写 try 代码块，但在 except 代码块中明确地告诉 Python 什么都不要做。 Python 有一个 pass 语句，可在代码块中使用它来让 Python什么都不要做

     ```python
     def count_words(filename):
     	""" 计算一个文件大致包含多少个单词 """
     	try:
     		--snip--
     	except FileNotFoundError:
     		pass
     	else:
     		--snip--
     filenames = ['1.txt', '2.txt', '3.txt']
     for filename in filenames:
     	count_words(filename)
         
     The file 1.txt has about 26 words.
     The file 2.txt has about 15 words.
     ```

     相比于前一个程序，这个程序唯一不同的地方是 pass 语句。现在，出现 FileNotFoundError 异常时，将执行 except 代码块中的代码，但什么都不会发生。这种错误发生时，不会出现 traceback ，也没有任何输出。用户将看到存在的每个文件包含多少个单词，但没有任何迹象表明有一个文件未找到

- 存储数据

  很多程序都要求用户输入某种信息，如让用户存储游戏首选项或提供要可视化的数据。不管专注的是什么，程序都把用户提供的信息存储在列表和字典等数据结构中。用户关闭程序时，你几乎总是要保存他们提供的信息；一种简单的方式是使用模块 json 来存储数据。模块 json 让你能够将简单的 Python 数据结构转储到文件中，并在程序再次运行时加载该文件中的数据。你还可以使用 json 在 Python 程序之间分享数据。更重要的是， JSON 数据格式并非 Python 专用的，这让你能够将以 JSON 格式存储的数据与使用其他编程语言的人分享。这是一种轻便格式，很有用，也易于学习。
  **注意** 　 JSON （ JavaScript Object Notation ）格式最初是为 JavaScript 开发的，但随后成了一种常见格式，被包括 Python 在内的众多语言采用

- 使用 json.dump()  和 json.load()

  我们来编写一个存储一组数字的简短程序，再编写一个将这些数字读取到内存中的程序。第一个程序将使用 json.dump() 来存储这组数字，而第二个程序将使用 json.load() 。函数 json.dump() 接受两个实参：要存储的数据以及可用于存储数据的文件对象。下面演示了如何使用 json.dump() 来存储数字列表

  ```python
  import json
  numbers = [2, 3, 5, 7, 11, 13]
  filename = 'numbers.json'
  with open(filename, 'w') as f_obj:
      json.dump(numbers, f_obj)
  ```

  这个程序没有输出，但我们可以打开文件 numbers.json ，看看其内容。数据的存储格式与 Python 中一样, 下面再编写一个程序，使用 json.load() 将这个列表读取到内存中：

  ```python
  import json
  filename = 'numbers.json'
  with open(filename) as f_obj:
      numbers = json.load(f_obj)
  print(numbers)
  
  [2, 3, 5, 7, 11, 13]
  ```

  这是一种在程序之间共享数据的简单方式

- 保存和读取用户生成的数据

  对于用户生成的数据，使用 json 保存它们大有裨益，因为如果不以某种方式进行存储，等程序停止运行时用户的信息将丢失。下面来看一个这样的例子：用户首次运行程序时被提示输入自己的名字，这样再次运行程序时就记住他了。我们先来存储用户的名字

  ```python
  import json
  username = input("What is your name? ")
  filename = 'username.json'
  with open(filename, 'w') as f_obj:
      json.dump(username, f_obj)  #调用 json.dump() ，并将用户名和一个文件对象传递给它，从而将用户名存储到文件中
      print("We'll remember you when you come back, " + username + "!")
      
  What is your name? mao
  We'll remember you when you come back, mao!
  ```

  现在再编写一个程序，向其名字被存储的用户发出问候

  ```python
  import json
  filename = 'username.json'
  with open(filename) as f_obj:
      username = json.load(f_obj)
      print("Welcome back, " + username + "!")
      
  Welcome back, mao!
  ```

  我们需要将这两个程序合并到一个程序（ remember_me.py ）中。这个程序运行时，我们将尝试从文件 username.json 中获取用户名，因此我们首先编写一个尝试恢复用户名的 try 代码块。如果这个文件不存在，我们就在 except 代码块中提示用户输入用户名，并将其存储在 username.json 中，以便程序再次运行时能够获取它

  ```python
  import json
  #  如果以前存储了用户名，就加载它
  # 否则，就提示用户输入用户名并存储它
  filename = 'username.json'
  try:
      with open(filename) as f_obj:   #尝试打开文件 username.json 。如果这个文件存在，就将其中的用户名读取到内存中
          username = json.load(f_obj)
  except FileNotFoundError:   #用户首次运行这个程序时，文件 username.json 不存在，将引发 FileNotFoundError 异常
      username = input("What is your name? ")
      with open(filename, 'w') as f_obj:
          json.dump(username, f_obj)
          print("We'll remember you when you come back, " + username + "!")
  else:
      print("Welcome back, " + username + "!")
      
  Welcome back, mao!
  ```

- 重构

  你经常会遇到这样的情况：代码能够正确地运行，但可做进一步的改进 —— 将代码划分为一系列完成具体工作的函数。这样的过程被称为 重构 。重构让代码更清晰、更易于理解、更容易扩展。要重构 remember_me.py ，可将其大部分逻辑放到一个或多个函数中。 remember_me.py 的重点是问候用户，因此我们将其所有代码都放到一个名为 greet_user() 的函数中

  ```python
  import json
  
  def get_stored_username():
      """ 如果存储了用户名，就获取它 """
      filename = 'username.json'
      """如果存储了用户名，这个函数就获取并返回它；如果文件 username.json 不存在，这个函数
      就返回 None"""
      try:
          with open(filename) as f_obj:
              username = json.load(f_obj)
      except FileNotFoundError:
          return None
      else:
          return username
  
  def get_new_username():
      """ 提示用户输入用户名 """
      username = input("What is your name? ")
      filename = 'username.json'
      with open(filename, 'w') as f_obj:
          json.dump(username, f_obj)
      return username
  
  def greet_user():
      """ 问候用户，并指出其名字 """
      username = get_stored_username()
      if username:
          print("Welcome back, " + username + "!")
      else:
          username = get_new_username()
          print("We'll remember you when you come back, " + username + "!")
  
  greet_user()
  ```

  在 remember_me.py 的这个最终版本中，每个函数都执行单一而清晰的任务。我们调用 greet_user() ，它打印一条合适的消息：要么欢迎老用户回来，要么问候新用户。为此，它首先调用 get_stored_username() ，这个函数只负责获取存储的用户名（如果存储了的话），再在必要时调用 get_new_username() ，这个函数只负责获取并存储新用户的用户名。要编写出清晰而易于维护和扩展的代码，这种划分工作必不可少。
